File: src/design_patterns/structural/composite/javascript/composite.js
// Component
class Component {
    operation() {
      throw new Error("operation() must be implemented");
    }
  }
  
  // Leaf
  class Leaf extends Component {
    operation() {
      return "Leaf";
    }
  }
  
  // Composite
  class Composite extends Component {
    constructor() {
      super();
      this._children = [];
    }
  
    add(component) {
      this._children.push(component);
    }
  
    remove(component) {
      const index = this._children.indexOf(component);
      if (index !== -1) {
        this._children.splice(index, 1);
      }
    }
  
    operation() {
      const results = [];
      for (let child of this._children) {
        results.push(child.operation());
      }
      return `Composite(${results.join("+")})`;
    }
  }
  
  // Usage
  const leaf1 = new Leaf();
  const leaf2 = new Leaf();
  const composite1 = new Composite();
  composite1.add(leaf1);
  composite1.add(leaf2);
  const composite2 = new Composite();
  composite2.add(composite1);
  console.log(composite2.operation());  // Output: Composite(Leaf+Leaf)
  

File: src/design_patterns/structural/composite/c#/composite.cs
// Component
abstract class Component
{
    public abstract string Operation();
}

// Leaf
class Leaf : Component
{
    public override string Operation()
    {
        return "Leaf";
    }
}

// Composite
class Composite : Component
{
    private List<Component> _children = new List<Component>();

    public void Add(Component component)
    {
        _children.Add(component);
    }

    public void Remove(Component component)
    {
        _children.Remove(component);
    }

    public override string Operation()
    {
        var results = new List<string>();
        foreach (var child in _children)
        {
            results.Add(child.Operation());
        }
        return $"Composite({string.Join("+", results)})";
    }
}

/*
// Usage
var leaf1 = new Leaf();
var leaf2 = new Leaf();
var composite1 = new Composite();
composite1.Add(leaf1);
composite1.Add(leaf2);
var composite2 = new Composite();
composite2.Add(composite1);
Console.WriteLine(composite2.Operation());  // Output: Composite(Leaf+Leaf)
*/
